### ISSUE TITLE
invert check for `clippy::try_err` and disable it inside of try blocks

### ISSUE THREAD

**Issue by yaahc** [rust-lang/rust-clippy#5757]
Right now clippy lints instances of returning bare `Err`'s via `?`

```rust
Err("foo")?; // clippy hates this

return Err("foo".into()); // clippy's preference
```

This perhaps not the best default going forward. Here are the upcoming concerns that make me prefer using `Err("")?`.

## Try Blocks

Once try blocks are stabilized `return Err` and `Err?` will start to mean two different things. A bare return will return to the function scope, where as `?` returns are scoped to the innermost try block. To resolve this we should just not apply this lint to expressions inside of try blocks.

## Error Return Traces and Try Trait interop

In future rust may have error return traces added to the Try trait where each `?` location passes a location to the inner error type to store so it can build up its own lightweight no_std compatible backtrace type that doesn't include any irrelevant frames. These frames would most likely be passed in by the implementation of the `Try` trait for `Result`. By returning via `return Err("".into());` instead of `?` we avoid this codepath and prevent the error return traces from being populated.

In order to ensure the best compatibility with current and future version of the `Try` trait we should prefer utilizing the core error propagation logic when propagating errors rather than recreating a low budget version of that logic manually.

**Comment by yaahc** [comment]
I am worried about how much churn this could cause, given that everyone presumably has `return Err("".into());` all throughout their code at clippy's behest and having that suddenly turn around and start yelling that it should be the other way will probably be frustrating.

We might want to plan out a transition plan between the two versions that minimizes churn / annoyance. It might be as simple as disabling this lint for now and re-enabling it later once more parts of the `Try` trait and try blocks have stabilized and it starts to matter which one you use again. Overall though I firmly believe that `Err(...)?` is the better of the two options and that clippy should ultimately have a lint that recommends this option whenever it can.

**Comment by joshtriplett** [comment]
I agree. I think this lint should just be disabled by default (and moved to `restriction`).

If, one day, we add a construct for throwing an error (e.g. `yeet e`), it'd make sense to have a lint steering people to that. But until then, I think we should allow `Err(e)?` as a widespread common pattern.

(I encountered this when running clippy on one of my own projects.)

**Comment by yaahc** [comment]
I went ahead and tagged this and am gonna submit it to TWIR for inclusion in the next issue. It's been a while since I worked on clippy though so if these tags look wrong lmk.

**Comment by nellorocca** [comment]
I'll take this one.
